File: 2512.reward-top-k-students.py
#
# @lc app=leetcode id=2512 lang=python3
#
# [2512] Reward Top K Students
#

# @lc code=start
from collections import defaultdict
import heapq


class Solution:
    def topStudents(
        self, pos: List[str], neg: List[str], report: List[str], student_id: List[int], k: int
    ) -> List[int]:

        pos, neg = set(pos), set(neg)

        res = []
        for id, report in zip(student_id, report):
            score = 0
            for word in report.split():
                if word in pos:
                    score += 3
                elif word in neg:
                    score -= 1
            res.append((-score, id))
        return (each[1] for each in heapq.nsmallest(k, res))


# @lc code=end

File: 2512.reward-top-k-students.rs
/*
 * @lc app=leetcode id=2512 lang=rust
 *
 * [2512] Reward Top K Students
 */

// @lc code=start
use std::collections::HashSet;

impl Solution {
    pub fn top_students(
        positive_feedback: Vec<String>,
        negative_feedback: Vec<String>,
        report: Vec<String>,
        student_id: Vec<i32>,
        k: i32,
    ) -> Vec<i32> {
        let pos: HashSet<_> = positive_feedback.into_iter().collect();
        let neg: HashSet<_> = negative_feedback.into_iter().collect();
        let mut res: Vec<_> = student_id
            .into_iter()
            .zip(report)
            .map(|(id, report)| {
                let mut points = 0;
                for word in report.split_ascii_whitespace() {
                    if pos.contains(word) {
                        points += 3;
                    } else if neg.contains(word) {
                        points -= 1;
                    }
                }
                (id, points)
            })
            .collect();

        res.sort_unstable_by(|a, b| b.1.cmp(&a.1).then(a.0.cmp(&b.0)));
        res.into_iter().take(k as usize).map(|(id, _)| id).collect()
    }
}
// @lc code=end
